### ISSUE TITLE
It should be possible to refer to images by reference

### ISSUE THREAD

**Issue by ctrlcctrlv** [fontforge/fontforge#3872]
My PNG feature is pretty good, (or at least I think so;) even with the low compression value I chose, I get 60&ndash;90% savings in SFD file sizes.

That's all well and good, but having to rewrite the files every time is inefficient, _obviously_. It was inefficient before when it was bitmap, now it's only less inefficient.

<hr/>

So, I want to make it possible to refer to images by reference. I want to do this with my new `Image2` tag. Right now FontForge only understands `Image2` with `image/png`, but we made `Image2` to be very extensible, so this is no problem. The MIME type is the magic key, and determines the parameters and their order.

I want to add a new MIME type, `inode/symlink`. The next argument is the _expected_ MIME type of the target file, so let's say `image/jpeg`, this informs what reader FontForge should try to use, and is needed because FontForge doesn't always have libpng or libjpeg or others (it also of course eases implementation, so we don't have to read magic bytes like `PNG`, `\xff\xd8\xff\xe0` etc.). Like for `image/png` type in the 20190801 release, we put the filesize in bytes to make implementation easier. For compatibility, we put a line break first, so that it makes sense to have an `EndImage2`. Then comes a hash, a [G_CHECKSUM_SHA256](https://developer.gnome.org/glib/stable/glib-Data-Checksums.html#GChecksumType), in hex format. Then comes another linebreak and the the absolute path of the file. 

Here's an example:

```
Image2: inode/symlink image/jpeg 316689
c65574b2a30a79a72365ca3859105599c667a8d4ca35d7e5d27f63b07a5294b1
/home/fred/Workspace/chomsky/canon/B.jpg
EndImage2
```

<hr>

**Does anyone reject my planned implementation?** Please speak now before I start working on it, or forever hold your peace, as they say. This is a blocker for a new (somewhat ambitious, if I'm honest) font of mine called TT2020 so I'd like to start working on it right away.

cc: @frank-trampe @jtanx Thanks for your time!

**Comment by jtanx** [comment]
So a couple of points:

* `inode/symlink` doesn't sound like an appropriate mime type to be using
* Why is the size, or even the hash needed, either the image is valid or not, and the file itself will encode the file size.
* Is this not missing the xoff/yoff/xscale/yscale components?

**Comment by ctrlcctrlv** [comment]
@jtanx:

* OK. I thought it was cute, but if you don't like it, how about `application/vnd.fontforge-file-reference`?
* For sanity and convenience. The size is useful for preparing a buffer to read from when we're reloading the file. The hash is useful for making sure that the file contains what we were able to read last time.
* Oh, you're right, it is. Let me add those in

**Comment by ctrlcctrlv** [comment]
I really would recommend keeping the hash so we can warn users about externally changed files, and having the image type will make things so much easier.

**Comment by jtanx** [comment]
> For sanity and convenience. The size is useful for preparing a buffer to read from when we're reloading the file. The hash is useful for making sure that the file contains what we were able to read last time.

I don't buy that, the GImage import code handles all that just fine without knowing any of that up-front. And on the hash, what are you going to do if the hash doesn't match? Discard it? Even if it's a valid image?

**Comment by ctrlcctrlv** [comment]
Nah, just stick a warning in the Warnings dialog. There are legitimate reasons for the image changing (something which was never possible before).

**Comment by jtanx** [comment]
In that case, I don't see too much point on the hash.

If you're talking about detecting externally changed files, once you've already loaded it, normally you do that through file notification methods (e.g. inotify, abstracted in glib/gio as https://developer.gnome.org/gio/stable/GFileMonitor.html#GFileMonitorEvent), which doesn't require hashing at all.

**Comment by ctrlcctrlv** [comment]
I'd really like FontForge to warn me if my images change as it could point to corruption, not all of which would invalidate files. FontForge warns about any number of things, including perfectly valid tables it just simply doesn't recognize. We already have libglib so I don't understand the resistance to a hash

**Comment by ctrlcctrlv** [comment]
(I'm only talking about on reloading SFD file, not inotify.)

**Comment by jtanx** [comment]
I guess, given that we're only using it to warn if the hash changes, I'm just questioning how useful this will actually be, vs. the associated cost of doing so. By adding hashing you most likely now have to read the file twice, once to compute the hash, and once more to load it in whatever image loader gets used.

I don't know, perhaps that's perfectly fine, and I can see some utility to it, but I also don't want to overcomplicate it if we don't have to.

**Comment by ctrlcctrlv** [comment]
Yeah, I feel like since this gives people brand new ways to shoot themselves in the foot, warnings are in order. Users might think they can just crop images in the folder, just as an example of something that might make sense to a user but will mess up FontForge's scaling.

Having hashes around is also good for VCS, and could down the road have more benefits.

By the way, mind commenting on my note with &dagger;? That's something else I've wanted to do for a while, but it's not specifically TT2020 related so I'm holding off on that until TT2020 is out.

**Comment by jtanx** [comment]
Regarding rotation, without deep diving into this, I don't have any strong opinions on this. I presume those xoff/yoff/xscale/yscale parameters already define some scaling params, so maybe you could just change it so it instead defines some sort of transformation matrix. Or perhaps just add a rotation value e.g. in degrees, if that's more intuitive. Not really something I've looked to deeply into in the past.

On your current proposal, I still hold that `image/jpeg 316689` is redundant information, which isn't required, the file size in particular. There's just no way you can use it, you can't presume that the file will actually be that size.

**Comment by ctrlcctrlv** [comment]
Regarding the filesize, I see your point and will drop that. I was planning to just compare them, but since you've accepted the hash I suppose we don't really need both.

What if the file was originally a JPEG but a PNG now exists in its place, or the reverse? Certainly the hash will help with this, but it'd be good to know information like this, I don't think we should load a JPEG if we expected a PNG. Some drawing programs, especially in proprietary operating systems, can silently change the types of files.

Another reason I'd like to keep the MIME around is for scripts that act on SFDs but which aren't FontForge. One can easily imagine @khaledhosny adding image validity checking to his excellent SFD normalizer (or not, this isn't a feature request, just an example scenario where the information is useful). More information would make that process easier.

**Comment by ctrlcctrlv** [comment]
(By the way, `application/vnd.fontforge-file-reference`, yay or nay? That's a MIME-compliant type, but I mean, we don't necessarily have to follow the MIME standard, I could just simply write `reference`, `Image2` doesn't use the MIME sub-elements in any way.)

**Comment by frank-trampe** [comment]
I don't think that it needs to be that specific if it is just a file path. We might encode to a URI and use "text/uri-list" instead. MIME types also support compounding with a plus sign, so, if we wanted to encode the MIME type of the pointed file (which I'm not necessarily advocating), we could do that in one string.

**Comment by ctrlcctrlv** [comment]
These are my proposed changes to `gimage.h`:

```diff
diff --git a/inc/gimage.h b/inc/gimage.h
index 5d7da3c00..4b1d0c954 100644
--- a/inc/gimage.h
+++ b/inc/gimage.h
@@ -86,6 +86,18 @@ typedef struct revcmap RevCMap;
 
 enum image_type { it_mono, it_bitmap=it_mono, it_index, it_true, it_rgba };
 
+enum image_format { if_png, if_bmp, if_gif, if_jpeg, if_xbm, if_tiff };
+
+// g_checksum_type_get_length(G_CHECKSUM_SHA256)
+#define SHA256_DIGEST_LEN 32
+// For images that are referred to by reference, not embedded in SFD.
+struct ReferenceData {
+    bool reference;
+    char* filename;
+    char hash[SHA256_DIGEST_LEN];
+    enum image_format format;
+};
+
 struct _GImage {
 /* Format: bitmaps are stored with the most significant bit first in byte units
 	    indexed    images are stored in byte units
@@ -100,6 +112,7 @@ struct _GImage {
     GClut *clut;
     Color trans;		/* PNG supports more than one transparent color, we don't */
 				/* for non-true color images this is the index, not a color */
+    struct ReferenceData refdata;
 };
 
 /* We deal with 1 bit, 8 bit and 32 bit images internal. 1 bit images may have*/
```

Thoughts?

**Comment by ctrlcctrlv** [comment]
By the way, @jtanx was right, the implementation already guesses MIME, and a hash warning is enough to prevent errant programs from overwriting PNGs as JPEGs.

So I will not be including a MIME type. Still working on this, hoping to open my PR by Wednesday.

**Comment by ctrlcctrlv** [comment]
I'm still working on this. Because I don't want to PR something half baked, I need a few more days, because I'm adding a way to link (which is another reason I need a hashing function) and unlink references.

Preliminary UI for this can be seen in the images in this tweet: https://twitter.com/FontForge/status/1166621838358831104
